File: android/src/test/java/de/idrinth/habitevaluator/android/StartupRegressionTest.java
package de.idrinth.habitevaluator.android;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import de.idrinth.habitevaluator.android.ui.ScreenPagerAdapter;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Regression test: verifies invariants that, if violated, would cause the app
 * to crash on startup or shortly after. These are static-state and constant
 * checks that can run without an Android context.
 */
class StartupRegressionTest {

    @BeforeEach
    void setUp() {
        // Reset mutable static state so tests are isolated
        MainActivity.setEditHabitId(null);
        MainActivity.setPointDevelopmentHabitId(null);
        MainActivity.setRecordEmotionPairId(null);
    }

    // ── Shared collection fields must never be null ─────────────────────────

    @Test
    void testSharedCategoriesIsNeverNull() {
        assertNotNull(
                MainActivity.getSharedCategories()
        ,
                "sharedCategories must be initialized to a non-null list to prevent NPE in fragments");
    }

    @Test
    void testSharedEmotionPairsIsNeverNull() {
        assertNotNull(
                MainActivity.getSharedEmotionPairs()
        ,
                "sharedEmotionPairs must be initialized to a non-null list to prevent NPE in fragments");
    }

    @Test
    void testSharedSleepEntriesIsNeverNull() {
        assertNotNull(
                MainActivity.getSharedSleepEntries()
        ,
                "sharedSleepEntries must be initialized to a non-null list to prevent NPE in fragments");
    }

    @Test
    void testSharedMedicationsIsNeverNull() {
        assertNotNull(
                MainActivity.getSharedMedications()
        ,
                "sharedMedications must be initialized to a non-null list to prevent NPE in fragments");
    }

    // ── Navigation IDs default to null (not empty or garbage) ───────────────

    @Test
    void testEditHabitIdDefaultsToNull() {
        assertNull(
                MainActivity.getEditHabitId()
        ,
                "editHabitId must default to null so navigation guards work correctly");
    }

    @Test
    void testPointDevelopmentHabitIdDefaultsToNull() {
        assertNull(
                MainActivity.getPointDevelopmentHabitId()
        ,
                "pointDevelopmentHabitId must default to null");
    }

    @Test
    void testRecordEmotionPairIdDefaultsToNull() {
        assertNull(
                MainActivity.getRecordEmotionPairId()
        ,
                "recordEmotionPairId must default to null");
    }

    // ── Remote storage defaults to off ──────────────────────────────────────

    @Test
    void testRemoteStorageDefaultsFalse() {
        assertFalse(
                MainActivity.isSharedUsingRemoteStorage()
        ,
                "Remote storage must default to false so the app can start without network");
    }

    // ── saveAllHabits must not throw when repository is null ────────────────

    @Test
    void testSaveAllHabitsDoesNotThrowWithNullRepository() {
        // Before onCreate runs, repositories are null. This method is called
        // from fragments and must handle that gracefully.
        MainActivity.saveAllHabits();
    }

    @Test
    void testRefreshSharedMedicationsDoesNotThrowWithNullRepository() {
        MainActivity.refreshSharedMedications();
    }

    // ── ScreenPagerAdapter page constants are valid ─────────────────────────

    @Test
    void testPageCountIsPositive() {
        assertTrue(
                ScreenPagerAdapter.PAGE_COUNT > 0
        ,
                "PAGE_COUNT must be positive");
    }

    @Test
    void testHomePageIsWithinPageCount() {
        assertTrue(
                ScreenPagerAdapter.PAGE_HOME >= 0
                        && ScreenPagerAdapter.PAGE_HOME < ScreenPagerAdapter.PAGE_COUNT
        ,
                "PAGE_HOME must be a valid index within PAGE_COUNT");
    }

    @Test
    void testAllNavigablePageIndicesAreWithinRange() {
        // Every page constant referenced by navigation must be within [0, PAGE_COUNT)
        int[] allPages = {
                ScreenPagerAdapter.PAGE_HOME,
                ScreenPagerAdapter.PAGE_DIARY,
                ScreenPagerAdapter.PAGE_SLEEP,
                ScreenPagerAdapter.PAGE_EMERGENCY_PLAN,
                ScreenPagerAdapter.PAGE_EMOTIONAL_STATE,
                ScreenPagerAdapter.PAGE_EDIT_HABITS,
                ScreenPagerAdapter.PAGE_ADD_HABIT,
                ScreenPagerAdapter.PAGE_STATS,
                ScreenPagerAdapter.PAGE_POINT_DEVELOPMENT,
                ScreenPagerAdapter.PAGE_ADD_EMOTION_PAIR,
                ScreenPagerAdapter.PAGE_RECORD_EMOTION_ENTRY,
                ScreenPagerAdapter.PAGE_SETTINGS,
                ScreenPagerAdapter.PAGE_IMPRINT,
                ScreenPagerAdapter.PAGE_POSITIVITY_DIARY,
                ScreenPagerAdapter.PAGE_SPORT_LOG,
                ScreenPagerAdapter.PAGE_FOOD_LOG,
                ScreenPagerAdapter.PAGE_MEDICATION_LOG,
                ScreenPagerAdapter.PAGE_MEDICATION_LIST,
                ScreenPagerAdapter.PAGE_ACTIVITY_LOG
        };

        for (int page : allPages) {
            assertTrue(
                    page >= 0 && page < ScreenPagerAdapter.PAGE_COUNT
            ,
                    "Page constant " + page + " must be >= 0 and < PAGE_COUNT ("
                            + ScreenPagerAdapter.PAGE_COUNT + ")");
        }
    }

    @Test
    void testPageCountEqualsNumberOfPageConstants() {
        // If someone adds a page constant but forgets to bump PAGE_COUNT,
        // createFragment will silently return a default HomeFragment
        // for that index — or worse, the ViewPager will crash.
        int[] allPages = {
                ScreenPagerAdapter.PAGE_HOME,
                ScreenPagerAdapter.PAGE_DIARY,
                ScreenPagerAdapter.PAGE_SLEEP,
                ScreenPagerAdapter.PAGE_EMERGENCY_PLAN,
                ScreenPagerAdapter.PAGE_EMOTIONAL_STATE,
                ScreenPagerAdapter.PAGE_EDIT_HABITS,
                ScreenPagerAdapter.PAGE_ADD_HABIT,
                ScreenPagerAdapter.PAGE_STATS,
                ScreenPagerAdapter.PAGE_POINT_DEVELOPMENT,
                ScreenPagerAdapter.PAGE_ADD_EMOTION_PAIR,
                ScreenPagerAdapter.PAGE_RECORD_EMOTION_ENTRY,
                ScreenPagerAdapter.PAGE_SETTINGS,
                ScreenPagerAdapter.PAGE_IMPRINT,
                ScreenPagerAdapter.PAGE_POSITIVITY_DIARY,
                ScreenPagerAdapter.PAGE_SPORT_LOG,
                ScreenPagerAdapter.PAGE_FOOD_LOG,
                ScreenPagerAdapter.PAGE_MEDICATION_LOG,
                ScreenPagerAdapter.PAGE_MEDICATION_LIST,
                ScreenPagerAdapter.PAGE_ACTIVITY_LOG
        };

        assertEquals(
                ScreenPagerAdapter.PAGE_COUNT, allPages.length
        ,
                "PAGE_COUNT must equal the total number of declared page constants");
    }

    @Test
    void testPageConstantsAreContiguousFromZero() {
        // The ViewPager2 iterates from 0..PAGE_COUNT-1, so pages must be
        // contiguous to avoid unmapped gaps that would hit the default case.
        int[] allPages = {
                ScreenPagerAdapter.PAGE_HOME,
                ScreenPagerAdapter.PAGE_DIARY,
                ScreenPagerAdapter.PAGE_SLEEP,
                ScreenPagerAdapter.PAGE_EMERGENCY_PLAN,
                ScreenPagerAdapter.PAGE_EMOTIONAL_STATE,
                ScreenPagerAdapter.PAGE_EDIT_HABITS,
                ScreenPagerAdapter.PAGE_ADD_HABIT,
                ScreenPagerAdapter.PAGE_STATS,
                ScreenPagerAdapter.PAGE_POINT_DEVELOPMENT,
                ScreenPagerAdapter.PAGE_ADD_EMOTION_PAIR,
                ScreenPagerAdapter.PAGE_RECORD_EMOTION_ENTRY,
                ScreenPagerAdapter.PAGE_SETTINGS,
                ScreenPagerAdapter.PAGE_IMPRINT,
                ScreenPagerAdapter.PAGE_POSITIVITY_DIARY,
                ScreenPagerAdapter.PAGE_SPORT_LOG,
                ScreenPagerAdapter.PAGE_FOOD_LOG,
                ScreenPagerAdapter.PAGE_MEDICATION_LOG,
                ScreenPagerAdapter.PAGE_MEDICATION_LIST,
                ScreenPagerAdapter.PAGE_ACTIVITY_LOG
        };

        boolean[] seen = new boolean[ScreenPagerAdapter.PAGE_COUNT];
        for (int page : allPages) {
            assertFalse( seen[page],"Duplicate page index: " + page);
            seen[page] = true;
        }
        for (int i = 0; i < seen.length; i++) {
            assertTrue( seen[i],"No page constant maps to index " + i);
        }
    }
}

File: android/src/test/kotlin/de/idrinth/habitevaluator/android/StartupRegressionTest.kt
package de.idrinth.habitevaluator.android

import de.idrinth.habitevaluator.android.ui.navigation.Screen
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*

/**
 * Regression test: verifies invariants that, if violated, would cause the app
 * to crash on startup or shortly after. In the Compose architecture, shared
 * mutable state lives in AppViewModel (StateFlow), and navigation uses
 * Screen sealed class routes instead of ViewPager2 page indices.
 */
class StartupRegressionTest {

    // ── Screen routes must be non-empty ─────────────────────────────────────

    @Test
    fun testHomeRouteIsNotEmpty() {
        assertTrue(Screen.Home.route.isNotEmpty(), "Home route must not be empty")
    }

    @Test
    fun testDiaryRouteIsNotEmpty() {
        assertTrue(Screen.Diary.route.isNotEmpty(), "Diary route must not be empty")
    }

    @Test
    fun testSleepRouteIsNotEmpty() {
        assertTrue(Screen.Sleep.route.isNotEmpty(), "Sleep route must not be empty")
    }

    @Test
    fun testEmergencyPlanRouteIsNotEmpty() {
        assertTrue(Screen.EmergencyPlan.route.isNotEmpty(), "EmergencyPlan route must not be empty")
    }

    @Test
    fun testEmotionalStateRouteIsNotEmpty() {
        assertTrue(Screen.EmotionalState.route.isNotEmpty(), "EmotionalState route must not be empty")
    }

    @Test
    fun testPlannerRouteIsNotEmpty() {
        assertTrue(Screen.Planner.route.isNotEmpty(), "Planner route must not be empty")
    }

    // ── All screen routes must be unique ────────────────────────────────────

    @Test
    fun testAllScreenRoutesAreUnique() {
        val allRoutes = listOf(
            Screen.Home.route,
            Screen.Diary.route,
            Screen.Sleep.route,
            Screen.EmergencyPlan.route,
            Screen.EmotionalState.route,
            Screen.EditHabits.route,
            Screen.AddHabit.route,
            Screen.Stats.route,
            Screen.PointDevelopment.route,
            Screen.AddEmotionPair.route,
            Screen.RecordEmotionEntry.route,
            Screen.Settings.route,
            Screen.Imprint.route,
            Screen.PositivityDiary.route,
            Screen.SportLog.route,
            Screen.FoodLog.route,
            Screen.MedicationLog.route,
            Screen.MedicationList.route,
            Screen.ActivityLog.route,
            Screen.PdfExport.route,
            Screen.SleepAnalysis.route,
            Screen.Correlations.route,
            Screen.EmergencyDialogue.route,
            Screen.Planner.route
        )

        val uniqueRoutes = allRoutes.toSet()
        assertEquals(
            allRoutes.size, uniqueRoutes.size,
            "All screen routes must be unique. Duplicates: ${allRoutes.groupBy { it }.filter { it.value.size > 1 }.keys}"
        )
    }

    @Test
    fun testAllScreenRoutesAreNonBlank() {
        val allRoutes = listOf(
            Screen.Home.route,
            Screen.Diary.route,
            Screen.Sleep.route,
            Screen.EmergencyPlan.route,
            Screen.EmotionalState.route,
            Screen.EditHabits.route,
            Screen.AddHabit.route,
            Screen.Stats.route,
            Screen.PointDevelopment.route,
            Screen.AddEmotionPair.route,
            Screen.RecordEmotionEntry.route,
            Screen.Settings.route,
            Screen.Imprint.route,
            Screen.PositivityDiary.route,
            Screen.SportLog.route,
            Screen.FoodLog.route,
            Screen.MedicationLog.route,
            Screen.MedicationList.route,
            Screen.ActivityLog.route,
            Screen.PdfExport.route,
            Screen.SleepAnalysis.route,
            Screen.Correlations.route,
            Screen.EmergencyDialogue.route,
            Screen.Planner.route
        )

        for (route in allRoutes) {
            assertTrue(route.isNotBlank(), "Screen route must not be blank: '$route'")
        }
    }

    // ── Parameterized routes generate correctly ─────────────────────────────

    @Test
    fun testEditHabitsCreateRouteContainsHabitId() {
        val route = Screen.EditHabits.createRoute("test-id")
        assertTrue(route.contains("test-id"), "EditHabits route must contain the habit ID")
    }

    @Test
    fun testPointDevelopmentCreateRouteContainsHabitId() {
        val route = Screen.PointDevelopment.createRoute("test-id")
        assertTrue(route.contains("test-id"), "PointDevelopment route must contain the habit ID")
    }

    @Test
    fun testRecordEmotionEntryCreateRouteContainsPairId() {
        val route = Screen.RecordEmotionEntry.createRoute("pair-id")
        assertTrue(route.contains("pair-id"), "RecordEmotionEntry route must contain the pair ID")
    }

    // ── AppViewModel class exists ───────────────────────────────────────────

    @Test
    fun testAppViewModelClassCanBeReferenced() {
        val clazz = AppViewModel::class.java
        assertNotNull(clazz)
    }

    @Test
    fun testAppViewModelClassNameIsCorrect() {
        assertEquals("AppViewModel", AppViewModel::class.java.simpleName)
    }

    // ── Screen count matches expected total ─────────────────────────────────

    @Test
    fun testTotalScreenCount() {
        val allRoutes = listOf(
            Screen.Home.route,
            Screen.Diary.route,
            Screen.Sleep.route,
            Screen.EmergencyPlan.route,
            Screen.EmotionalState.route,
            Screen.EditHabits.route,
            Screen.AddHabit.route,
            Screen.Stats.route,
            Screen.PointDevelopment.route,
            Screen.AddEmotionPair.route,
            Screen.RecordEmotionEntry.route,
            Screen.Settings.route,
            Screen.Imprint.route,
            Screen.PositivityDiary.route,
            Screen.SportLog.route,
            Screen.FoodLog.route,
            Screen.MedicationLog.route,
            Screen.MedicationList.route,
            Screen.ActivityLog.route,
            Screen.PdfExport.route,
            Screen.SleepAnalysis.route,
            Screen.Correlations.route,
            Screen.EmergencyDialogue.route,
            Screen.Planner.route
        )
        assertEquals(24, allRoutes.size, "Total number of screen routes must be 24")
    }

    // ── MainActivity class exists ───────────────────────────────────────────

    @Test
    fun testMainActivityClassCanBeReferenced() {
        val clazz = MainActivity::class.java
        assertNotNull(clazz)
    }
}
